lib/msf/base/sessions/command_shell.rb

```diff
@@ -193,12 +196,28 @@ def shell_write(buf)
   # :category: Msf::Session::Provider::SingleCommandShell implementors
   #
   # Closes the shell.
+  # Note: parent's 'self.kill' method calls cleanup below.
   #
   def shell_close()
-    rstream.close rescue nil
     self.kill
   end
 
+  ##
+  # :category: Msf::Session implementors
+  #
+  # Closes the shell.
+  #
+  def cleanup
+    if rstream
+      if !@cleanup_command.blank?
+        shell_command_token(@cleanup_command, 0)
```

Comment:
A funny thing we noticed about shell_command_token is it always waits for the timeout before returning, even if the token is returned. Feels like a bug to me, since what's the point of waiting the full timeout if you're reading the token in the first place?

Reply:
Have noticed this as well. Definitely wrong behavior.

Reply:
OK, I'll take a stab at fixing it while here.

Reply:
It works with https://github.com/rapid7/metasploit-framework/pull/9356

I didn't bother trying to untangle the ringbuffer code, since it worked without it, and makes the Meterpreter shell and command shell paths identical.